Implementation to Provide RGB Grid Values for sigma_a and sigma_s

shadeops opened the pull request:
_**Note**: The is PR is a Draft of an implementation, if the idea makes sense for pbrt-v4 (especially at this late stage) then I'm happy to tidy up the code, otherwise we can reject it._

# Problem Overview

Uniform Grid's `rgb density.rgb` option is a difficult interface for specifying coloured volumetrics. This is due to rgb values being used both for sigma_a and sigma_s.
https://github.com/mmp/pbrt-v4/blob/77d367204ad7c73eb5072c62f5e6f12604f0574d/src/pbrt/media.h#L386-L391

An rgb density value of (1,0,0) means you'll be scattering and absorbing red light in the cross section. The absorption of the red light results in a cyan being transmitted.

### Examples
The following are some renders of a density.rgb  setup, where the rgb values of each smoke plume start at red on the left and go purple on the right. In other words, the left most plume is red, the one next to it orange, etc.

#### Example 1: 
- sigma_a = {1,1,1};
- sigma_s = {1,1,1};
![rainbow_density_rgb](https://user-images.githubusercontent.com/2104033/135335695-0084402a-1092-4754-8b41-22862af3495a.png)

#### Example 2: 
- sigma_a = {0,0,0};
- sigma_s = {1,1,1};
![rainbow_density_rgb_a0](https://user-images.githubusercontent.com/2104033/135349739-c14d40c2-3f77-40fb-9944-f2e0825dcb80.png)


- #### Example 3: 
- sigma_a = {1,1,1};
- sigma_s = {0,0,0};
![rainbow_density_rgb_s0](https://user-images.githubusercontent.com/2104033/135350032-a7b0c580-0dc0-4de3-b919-82b48a4a6d02.png)


# PR Changes
### Allow for "rgb" versions of sigma_a and sigma_s
Here we add two new parameters `rgb density.sigma_a.rgb` and `rgb density.sigma_s.rgb`. This gives us independent color controls over for our sigma values.

Additionally the user has the option to supply a float or rgb version of sigma_a / sigma_s.

#### Example: 
Using the same example as before but instead of specifying `density.rgb` we use `density.sigma_a.rgb` and `density.sigma_s.rgb`. (The values in for sigma_a.rgb are the complement of sigma_s.rgb)

![rainbow](https://user-images.githubusercontent.com/2104033/135336332-f4802cdb-1c65-4651-92ef-f7f5b85b3da4.png)
 
### Evaluation of Changes
#### Pros
- More control and better modelling.

#### Cons
- Increased code complexity. While the amount of additional rendering code is minimal, there is a lot of extra error checking that happens now because of the number of valid parameter combinations goes from 3 to 6. ie)
  - `float density`
  - `rgb density.rgb`
  - `float density.sigma_a` and `float density.sigma_s`
  - `float density.sigma_a` and `rgb density.sigma_s.rgb`
  - `rgb density.sigma_a.rgb` and `float density.sigma_s`
  - `rgb density.sigma_a.rgb` and `rgb density.sigma_s.rgb`

mmp commented:
Yep, this is worth doing (and the images are compelling.) On the plus side, most of the error checking doesn't appear in the book text, so this would cause relatively little disruption in return for what it gives.

I am wondering if it makes sense to break up the `UniformGridMediumProvider` into multiple classes and then have `Medium::Create()` create the appropriate type based on which parameters were actually provided. (That would be a bit more disruptive to the text, but not (yet) the end of the world.)

shadeops commented:
Updated the PR and removed the conditional salad for validating the parameters. Should be more compact, readable and helpful.

Regarding splitting up UniformGridMediumProvider into multiple classes, I was wondering if the extra conditional logic in UniformGridMediumProvider::Density would have a noticeable impact on render times. While I haven't tested comprehensively a few quick tests between the current approach and a specialized ::Density() didn't show any difference in render times. Hooray for branch prediction and other arcane magic that modern CPU/GPUs implement?

mmp commented:
On multiple classes, I'm more thinking about complexity in describing the code in the book (where lots of conditionals make things messier) than performance. After some more thought, what do you think of something like this:

"grid" medium -> `UniformGridMediumProvider`. Takes sigma_a and sigma_s as spectral quantities and then a grid of density values. sigma_a/sigma_s are computed inside the medium by scaling the baseline spectra by the trilinearly interpolated density (as now). Emission can be specified via a grid of blackbody temperatures and a grid of emission scaling factors.

"rgbgrid" medium -> `UniformRGBGridMediumProvider` (ugh--long name.. Maybe remove `Uniform` from both class names?). Takes a grid of RGB values for sigma_a, another RGB grid for sigma_s. Both are optional, in which case 0 is assumed. Then, optionally, a third grid of RGB for emission values (in which case sigma_a is required).

Then the object creation code isn't choosing a class type based on the parameters--each grid type has a set of required parameters and a set of optional ones. Although there are some combinations that are supported now (e.g. spectral sigma_a/sigma_s * density but RGB emission), IMHO those two cover the majority of the important use cases. (Though, having not gotten all the use cases right before, I may be wrong about that...)

shadeops commented:
> Maybe remove Uniform from both class names?

I wouldn't object to this. Generally when I hear people refer to the data that comprises mediums/volumes it tends to be just "grids" or "fields". I have to unpretzel my brain at times because in DCCs a ["uniform volume"](https://www.sidefx.com/docs/houdini/gallery/shop/vopmaterial/uniformvolume.html) is analogous to a HomogeneousMedium. Of course _uniform grid_ and _uniform volume_ mean two different things, but I have on occasion confused myself.


> "rgbgrid" medium -> UniformRGBGridMediumProvider (ugh--long name.. Maybe remove Uniform from both class names?). Takes a grid of RGB values for sigma_a, another RGB grid for sigma_s. Both are optional, in which case 0 is assumed. Then, optionally, a third grid of RGB for emission values (in which case sigma_a is required).

Ya, I like this, the "rgbgrid" becomes full control mode. Losing the float grid versions of sigma_a / sigma_s and keeping only the "rgb" is fine. Keeps the interface/code simple.

> "grid" medium -> UniformGridMediumProvider. Takes sigma_a and sigma_s as spectral quantities and then a grid of density values. sigma_a/sigma_s are computed inside the medium by scaling the baseline spectra by the trilinearly interpolated density (as now). Emission can be specified via a grid of blackbody temperatures and a grid of emission scaling factors.

This make sense, however for emission are you thinking of matching NanoVDBMediumProvider interface? 
##### NanoVDBMediumProvider Emission Parameters
- float grid temperature
- float LeScale
- float temperaturecutoff
- float temperaturescale 

Or for the "grid" medium you have a Lescale grid as well?

shadeops commented:
Thinking about the emission options for the GridMediumProvider.

## Option 1: Keep Current Interface
#### Emission Based Parameters
- `spectrum Le`
- `float Lescale` (grid)
#### Pros-
- Simple interface
- All rgb / spectrum are derived from non-grid parameters. Grid parameters just provide scaling.
#### Cons -
- ???
## Option 2: Match NanoVDBMediumProvider
#### Emission Based Parameters
- `float temperature` (grid)
- `float LeScale`
- `float temperaturecutoff`
- `float temperaturescale`
#### Pros-
- Allows more physical based emission interface for users who don't have access to NanoVDB.
- "nanovdb" and "grid" have the same interface, but demonstrate different approaches to providing grid data.
#### Cons -
- For making a simple emissive medium is more difficult than Option 1, as  you are constrained by blackbody radiation.
- Requires 
## Option 3: Suggested Interface?
(It is entirely possible I'm misunderstanding)
#### Emission Based Parameters
- `float temperature` (grid)
- `float Lescale` (grid)
- `float temperaturecutoff` ?
- `float temperaturescale` ?
#### Pros -
- Additional control over what NanoVDBMediumProvider provides.
#### Cons -
- More complex than Option 1, no way to make a simple non-blackbody based emissive medium without resorting to "rgbcolor".
- ???

mmp commented:
First, regarding "temperaturecutoff" and "temperaturescale" in the NanoVDB provider, that was basically just a hack to get some of the example OpenVDB volumes rendering nicely. IIRC they had temperature values in the range 0-40, so that was just to be able to scale them into more of a reasonable blackbody emission temperature range and then take it from there. For pbrt's "native" media specification formats, IMHO it's fine to say "this is how temperature is interpreted and now over to you, whoever is generating the scene description, to deal with it..."  (Otherwise we start to go down the road of specifying a full little expression language of offsets and scales and clamps for every value.)

Good point about the value of non-blackbody emission spectra for the not-RGB option... And as a general principle, it should somehow be possible to specify an arbitrary spectral emission profile, which isn't possible if the only options are blackbody and RGB.

So... What about if the non-RGB grid lets emission be specified either via:

- a specified spectrum that is optionally scaled by a specified grid of emission scale factors
- a grid of blackbody emission temperatures that are optionally scaled by that same grid of emission scale factors

?

shadeops commented:
> For pbrt's "native" media specification formats, IMHO it's fine to say "this is how temperature is interpreted and now over to you, whoever is generating the scene description, to deal with it

:+1:  Works for me!

> So... What about if the non-RGB grid lets emission be specified either via:
> 
>     a specified spectrum that is optionally scaled by a specified grid of emission scale factors
>     a grid of blackbody emission temperatures that are optionally scaled by that same grid of emission scale factors

Yea that sounds good, gives options without being overly complicated.

shadeops commented:
I took at stab at implementing the above (updating this PR).

Two things of note for the "rgbgrid"
- I added a `float LeScale` parameter to globally scale down the emissive "Le" grid.
- If the sigma_a or sigma_s grids are not provided, instead of assuming 0, I changed this to 1. Which I believe is more in line with the Cuboid Medium's default sigma_a/sigma_s. Basically not providing a sigma_s grid would result in a black medium which I think is counter to expectations.

Also attached some test scenes -
[examples.zip](https://github.com/mmp/pbrt-v4/files/7274785/examples.zip)

mmp commented on src/pbrt/media.h:
```diff
@@ -346,50 +346,47 @@ class CuboidMedium {
         Point3f pp = Point3f(bounds.Offset(p));
+        if (temperatureGrid) {
+            Float temp = temperatureGrid->Lookup(pp);
+            if (temp <= 100.f)
```
Out of curiosity, why this check?

shadeops replied:
Good question, I was attempting to match what  NanoVDBMediumProvider::Le() did.
https://github.com/mmp/pbrt-v4/blob/77d367204ad7c73eb5072c62f5e6f12604f0574d/src/pbrt/media.h#L625-L626

mmp replied:
Got it. I'd suggest that should be in the same category of "up to the user if they want that or not, since they're generating the input," but in any case let's get this landed and I'll revisit that in a bit...

mmp commented:
Many thanks for digging into this and for all of the iterations on the specifics of the design!

shadeops commented:
My pleasure! Now to render some more images. :smile:
